homeassistant/helpers/restore_state.py

```diff
@@ -1,98 +1,113 @@
-    _LOGGER.debug('Created cache with %s', list(hass.data[DATA_RESTORE_CACHE]))
+async def async_get_restore_cache(hass: HomeAssistant) -> Dict[str, State]:
+    """Get the restore cache for loading previous states."""
+    task = hass.data.get(DATA_RESTORE_CACHE_TASK)
 
+    if task is None:
+        async def _load_restore_cache(hass: HomeAssistant) -> Dict[str, State]:
+            """Load the restore cache to be used by other components."""
+            store = _get_restore_state_store(hass)
 
-@bind_hass
-async def async_get_last_state(hass, entity_id: str):
-    """Restore state."""
-    if DATA_RESTORE_CACHE in hass.data:
-        return hass.data[DATA_RESTORE_CACHE].get(entity_id)
+            states = await store.async_load()
+            if states is None:
+                _LOGGER.debug('Not creating cache - no saved states found')
+                return {}
 
-    if _RECORDER not in hass.config.components:
-        return None
+            cache = {
+                state['entity_id']: State.from_dict(state) for state in states}
+            _LOGGER.debug('Created cache with %s', list(cache))
 
-    if hass.state not in (CoreState.starting, CoreState.not_running):
-        _LOGGER.debug("Cache for %s can only be loaded during startup, not %s",
-                      entity_id, hass.state)
-        return None
+            return cache
 
-    try:
-        with async_timeout.timeout(RECORDER_TIMEOUT, loop=hass.loop):
-            connected = await wait_connection_ready(hass)
-    except asyncio.TimeoutError:
-        return None
+        task = hass.data[DATA_RESTORE_CACHE_TASK] = hass.async_create_task(
+            _load_restore_cache(hass))
 
-    if not connected:
-        return None
+    return await task
 
-    if _LOCK not in hass.data:
-        hass.data[_LOCK] = asyncio.Lock(loop=hass.loop)
 
-    async with hass.data[_LOCK]:
-        if DATA_RESTORE_CACHE not in hass.data:
-            await hass.async_add_job(
-                _load_restore_cache, hass)
+def _get_restore_state_store(hass: HomeAssistant) -> Store:
+    """Return the data store for the last session's states."""
+    if DATA_RESTORE_STORAGE not in hass.data:
+        hass.data[DATA_RESTORE_STORAGE] = hass.helpers.storage.Store(
+            STORAGE_VERSION, STORAGE_KEY, encoder=JSONEncoder)
 
-    return hass.data.get(DATA_RESTORE_CACHE, {}).get(entity_id)
+    return hass.data[DATA_RESTORE_STORAGE]
 
 
-async def async_restore_state(entity, extract_info):
-    """Call entity.async_restore_state with cached info."""
-    if entity.hass.state not in (CoreState.starting, CoreState.not_running):
-        _LOGGER.debug("Not restoring state for %s: Hass is not starting: %s",
-                      entity.entity_id, entity.hass.state)
-        return
+@bind_hass
+async def async_get_last_state(
+        hass: HomeAssistant, entity_id: str) -> Optional[State]:
+    """Restore state."""
+    if hass.state not in (CoreState.starting, CoreState.not_running):
```

Comment:
I don't remember why we did this, it seems like we will lose automation states if we reload while running?